sdk/nodejs/runtime/resource.ts

```diff
@@ -155,25 +185,36 @@ export function registerResource(res: Resource, t: string, name: string, custom:
         // Now run the operation, serializing the invocation if necessary.
         const opLabel = `monitor.registerResource(${label})`;
         runAsyncResourceOp(opLabel, async () => {
-            const resp: any = await debuggablePromise(new Promise((resolve, reject) =>
-                monitor.registerResource(req, (err: grpc.ServiceError, innerResponse: any) => {
-                    log.debug(`RegisterResource RPC finished: ${label}; err: ${err}, resp: ${innerResponse}`);
-                    if (err) {
-                        // If the monitor is unavailable, it is in the process of shutting down or has already
-                        // shut down. Don't emit an error and don't do any more RPCs, just exit.
-                        if (err.code === grpc.status.UNAVAILABLE) {
-                            log.debug("Resource monitor is terminating");
-                            process.exit(0);
+            let resp: any;
+            if (monitor) {
+                // If we're running with an attachment to the engine, perform the operation.
+                resp = await debuggablePromise(new Promise((resolve, reject) =>
+                    (monitor as any).registerResource(req, (err: grpc.ServiceError, innerResponse: any) => {
+                        log.debug(`RegisterResource RPC finished: ${label}; err: ${err}, resp: ${innerResponse}`);
+                        if (err) {
+                            // If the monitor is unavailable, it is in the process of shutting down or has already
+                            // shut down. Don't emit an error and don't do any more RPCs, just exit.
+                            if (err.code === grpc.status.UNAVAILABLE) {
+                                log.debug("Resource monitor is terminating");
+                                process.exit(0);
+                            }
+
+                            // Node lets us hack the message as long as we do it before accessing the `stack` property.
+                            preallocError.message = `failed to register new resource ${name} [${t}]: ${err.message}`;
+                            reject(preallocError);
                         }
-
-                        // Node lets us hack the message as long as we do it before accessing the `stack` property.
-                        preallocError.message = `failed to register new resource ${name} [${t}]: ${err.message}`;
-                        reject(preallocError);
-                    }
-                    else {
-                        resolve(innerResponse);
-                    }
-                })), opLabel);
+                        else {
+                            resolve(innerResponse);
+                        }
+                    })), opLabel);
+            } else {
+                // If we aren't attached to the engine, mock up a fake response for testing purposes.
+                resp = {
+                    getUrn: () => createUrn(t, name),
```

Comment:
maybe name (or document) 'createUrn' in a way that makes it clear that it's for testing/mocking purposes.